sql/bloom_filters.h

```diff
@@ -177,9 +184,153 @@ struct PatternedSimdBloomFilter
      basically, unnoticeable, well below the noise level */
 #endif
 
+#ifdef NEON_IMPLEMENTATION
+  uint64x2_t CalcHash(uint64x2_t vecData)
+  {
+    static constexpr uint64_t prime_mx2= 0x9FB21C651E98DF25ULL;
+    static constexpr uint64_t bitflip= 0xC73AB174C5ECD5A2ULL;
+    uint64x2_t step1= veorq_u64(vecData, vdupq_n_u64(bitflip));
+    uint64x2_t step2= veorq_u64(vshrq_n_u64(step1, 48), vshlq_n_u64(step1, 16));
+    uint64x2_t step3= veorq_u64(vshrq_n_u64(step1, 24), vshlq_n_u64(step1, 40));
+    uint64x2_t step4= veorq_u64(step1, veorq_u64(step2, step3));
+    uint64x2_t step5;
+    step5= vsetq_lane_u64(vgetq_lane_u64(step4, 0) * prime_mx2, step4, 0);
+    step5= vsetq_lane_u64(vgetq_lane_u64(step4, 1) * prime_mx2, step5, 1);
+    uint64x2_t step6= vshrq_n_u64(step5, 35);
+    uint64x2_t step7= vaddq_u64(step6, vdupq_n_u64(8));
+    uint64x2_t step8= veorq_u64(step5, step7);
+    uint64x2_t step9;
+    step9= vsetq_lane_u64(vgetq_lane_u64(step8, 0) * prime_mx2, step8, 0);
+    step9= vsetq_lane_u64(vgetq_lane_u64(step8, 1) * prime_mx2, step9, 1);
+    return veorq_u64(step9, vshrq_n_u64(step9, 28));
+  }
+
+  uint64x2_t GetBlockIdx(uint64x2_t vecHash)
+  {
+    uint64x2_t vecNumBlocksMask= vdupq_n_u64(num_blocks - 1);
+    uint64x2_t vecBlockIdx= vshrq_n_u64(vecHash, mask_idx_bits + rotate_bits);
+    return vandq_u64(vecBlockIdx, vecNumBlocksMask);
+  }
+
+  uint64x2_t ConstructMask(uint64x2_t vecHash)
+  {
+    uint64x2_t vecMaskIdxMask= vdupq_n_u64((1 << mask_idx_bits) - 1);
+    uint64x2_t vecMaskMask= vdupq_n_u64((1ull << bits_per_mask) - 1);
+
+    uint64x2_t vecMaskIdx= vandq_u64(vecHash, vecMaskIdxMask);
+    uint64x2_t vecMaskByteIdx= vshrq_n_u64(vecMaskIdx, 3);
+    int64x2_t vecMaskBitIdx=
+      vsubq_s64(vdupq_n_s64(0),
+                vreinterpretq_s64_u64(vandq_u64(vecMaskIdx, vdupq_n_u64(0x7))));
+    uint64x2_t vecRawMasks= vdupq_n_u64(*reinterpret_cast<const uint64_t*>
+                              (masks + vgetq_lane_u64(vecMaskByteIdx, 0)));
+    vecRawMasks= vsetq_lane_u64(*reinterpret_cast<const uint64_t*>
+                   (masks + vgetq_lane_u64(vecMaskByteIdx, 1)), vecRawMasks, 1);
```

Comment:
hmm, does this neon specific implementation even make sense? only for two 64-bit values at a time, multiplication of array lookups are done per value, not simd. Perhaps a default implementation of the bloom filter will be just as fast? It'd make the code simpler.

Reply:
I did ask myself the same question in this commit comment:
> Performance improvement (microbenchmark) for bloom filters is less exciting,
> within 10-30% ballpark depending on compiler options and load.

I could try shuffling intrinsics to gain better improvement if you feel it is worthy (considering how BF are used and if they're bottleneck anyway). Generally I tend to agree we should put it on a shelve, unless you see good reason to have it.

Reply:
10-30% isn't that bad and the change is highly local, one header file only. may be it's worth it after all